src/box/lua/call.c

```diff
@@ -93,74 +93,84 @@ get_call_serializer(void)
+ * Returns number of items pushed (1 or 2) or -1 in case of error (diag is set).
  */
 static int
 box_lua_find(lua_State *L, const char *name, const char *name_end)
 {
-	int index = LUA_GLOBALSINDEX;
-	int objstack = 0, top = lua_gettop(L);
-	const char *start = name, *end;
-
-	while ((end = (const char *) memchr(start, '.', name_end - start))) {
-		lua_checkstack(L, 3);
-		lua_pushlstring(L, start, end - start);
-		lua_gettable(L, index);
-		if (! lua_istable(L, -1)) {
-			diag_set(ClientError, ER_NO_SUCH_PROC,
-				 name_end - name, name);
-			return -1;
-		}
-		start = end + 1; /* next piece of a.b.c */
-		index = lua_gettop(L); /* top of the stack */
-	}
+	lua_checkstack(L, 2); /* No more than 2 entries are needed. */
+	int top = lua_gettop(L);
 
-	/* box.something:method */
-	if ((end = (const char *) memchr(start, ':', name_end - start))) {
-		lua_checkstack(L, 3);
-		lua_pushlstring(L, start, end - start);
-		lua_gettable(L, index);
-		if (! (lua_istable(L, -1) ||
-			lua_islightuserdata(L, -1) || lua_isuserdata(L, -1) )) {
-				diag_set(ClientError, ER_NO_SUCH_PROC,
-					  name_end - name, name);
-				return -1;
+	/* Take the first token. */
+	const char *start = name;
+	while (start != name_end && strchr(".:[", *start) == NULL)
+		start++;
+	lua_pushlstring(L, name, start - name);
+	lua_gettable(L, LUA_GLOBALSINDEX);
+
+	/* Take the rest tokens. */
+	while (start != name_end) {
+		if (!lua_istable(L, -1) &&
+		    !lua_islightuserdata(L, -1) && !lua_isuserdata(L, -1))
+			goto no_such_proc;
+
+		char delim = *start++; /* skip delimiter. */
+		if (delim == '.') {
+			/* Look for the next token. */
+			const char *end = start;
+			while (end != name_end && strchr(".:[", *end) == NULL)
+				end++;
+			lua_pushlstring(L, start, end - start);
+			start = end;
+		} else if (delim == ':') {
+			lua_pushlstring(L, start, name_end - start);
+			lua_gettable(L, -2); /* get function from object. */
+			lua_insert(L, -2); /* swap function and object. */
+			break;
+		} else {
+			assert(delim == '[');
+			const char *end = memchr(start, ']', name_end - start);
+			if (end == NULL ||
+			    (end != name_end && strchr(".:[", end[1]) == NULL))
+				goto no_such_proc;
+
+			if (end - start >= 2 && start[0] == end[-1] &&
+			    (start[0] == '"' || start[0] == '\'')) {
+				/* Quoted string, just extract it. */
+				lua_pushlstring(L, start + 1, end - start - 2);
+			} else {
+				/* Must be a number, convert from string. */
+				lua_pushlstring(L, start, end - start);
+				int success;
+				lua_Number num = lua_tonumberx(L, -1, &success);
+				if (!success)
+					goto no_such_proc;
+				lua_pop(L, 1);
+				lua_pushnumber(L, num);
+			}
+			start = end + 1; /* skip closing bracket. */
 		}
 
-		start = end + 1; /* next piece of a.b.c */
-		index = lua_gettop(L); /* top of the stack */
-		objstack = index - top;
+		lua_gettable(L, -2); /* get child object from parent object. */
+		lua_remove(L, -2); /* drop previous parent object. */
```

Comment:
I have a concern your stack manipulation is not optimal: I think the original implementation used stack indices for a reason.

Reply:
I terms of stack I make N operations with O(1) cost here, while the original code makes one O(N) operation.  Perhaps we could save the time of foreign call itself, but the original code has to call `lua_checkstack` each round, so I bet it's a fair trade.